Core/Dom/XmlDomExtensions.cs

```diff
@@ -208,5 +208,34 @@ public static IEnumerable<T> SelfAndParentsOfType<T> (this XObject obj)
 				obj = obj.Parent;
 			}
 		}
+
+		public static XNode GetNodeContainingRange (this XNode node, TextSpan span)
+		{
+			if (node is XContainer container) {
+				foreach (var child in container.Nodes) {
+					var found = child.GetNodeContainingRange (span);
+					if (found != null) {
+						return found;
+					}
+				}
+			}
+
+			if (node.Span.Contains (span)) {
+				return node;
+			}
+
+			return null;
+		}
+
+
+		public static void VisitSelfAndChildren (this XNode node, Action<XNode> action)
```

Comment:
IMO this should be `VisitSelfAndDescendents`

Reply:
TIL descendants vs. descendents. Seeing that you're using descendents everywhere I will use that too